{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Class to store parameters of undirected graphical model of\n",
    "sequences and perform calculations using the model\n",
    "(statistical energies, coupling scores).\n",
    "Authors:\n",
    "  Thomas A. Hopf\n",
    "\"\"\"\n",
    "from collections import Iterable\n",
    "from copy import deepcopy\n",
    "\n",
    "from numba import jit\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Constants\n",
    "\n",
    "_SLICE = np.s_[:]\n",
    "HAMILTONIAN_COMPONENTS = [FULL, COUPLINGS, FIELDS] = [0, 1, 2]\n",
    "NUM_COMPONENTS = len(HAMILTONIAN_COMPONENTS)\n",
    "\n",
    "# Methods for fast calculations (moved outside of class for numba jit)\n",
    "\n",
    "\n",
    "@jit(nopython=True)\n",
    "def _hamiltonians(sequences, J_ij, h_i):\n",
    "    \"\"\"\n",
    "    Calculates the Hamiltonian of the global probability distribution P(A_1, ..., A_L)\n",
    "    for a given sequence A_1,...,A_L from J_ij and h_i parameters\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequences : np.array\n",
    "        Sequence matrix for which Hamiltonians will be computed\n",
    "    J_ij: np.array\n",
    "        L x L x num_symbols x num_symbols J_ij pair coupling parameter matrix\n",
    "    h_i: np.array\n",
    "        L x num_symbols h_i fields parameter matrix\n",
    "    Returns\n",
    "    -------\n",
    "    np.array\n",
    "        Float matrix of size len(sequences) x 3, where each row corresponds to the\n",
    "        1) total Hamiltonian of sequence and the 2) J_ij and 3) h_i sub-sums\n",
    "    \"\"\"\n",
    "    # iterate over sequences\n",
    "    N, L = sequences.shape\n",
    "    H = np.zeros((N, NUM_COMPONENTS))\n",
    "    for s in range(N):\n",
    "        A = sequences[s]\n",
    "        hi_sum = 0.0\n",
    "        Jij_sum = 0.0\n",
    "        for i in range(L):\n",
    "            hi_sum += h_i[i, A[i]]\n",
    "            for j in range(i + 1, L):\n",
    "                Jij_sum += J_ij[i, j, A[i], A[j]]\n",
    "\n",
    "        H[s] = [Jij_sum + hi_sum, Jij_sum, hi_sum]\n",
    "\n",
    "    return H\n",
    "\n",
    "\n",
    "@jit(nopython=True)\n",
    "def _single_mutant_hamiltonians(target_seq, J_ij, h_i):\n",
    "    \"\"\"\n",
    "    Calculate matrix of all possible single-site substitutions\n",
    "    Parameters\n",
    "    ----------\n",
    "    L : int\n",
    "        Length of model\n",
    "    num_symbols : int\n",
    "        Number of states of model\n",
    "    target_seq : np.array(int)\n",
    "        Target sequence for which mutant energy differences will be calculated\n",
    "    J_ij: np.array\n",
    "        L x L x num_symbols x num_symbols J_ij pair coupling parameter matrix\n",
    "    h_i: np.array\n",
    "        L x num_symbols h_i fields parameter matrix\n",
    "    Returns\n",
    "    -------\n",
    "    np.array\n",
    "        Float matrix of size L x num_symbols x 3, where the first two dimensions correspond to\n",
    "        Hamiltonian differences compared to target sequence for all possible substitutions in\n",
    "        all positions, and the third dimension corresponds to the deltas of\n",
    "        1) total Hamiltonian and the 2) J_ij and 3) h_i sub-sums\n",
    "    \"\"\"\n",
    "    L, num_symbols = h_i.shape\n",
    "    H = np.empty((L, num_symbols, NUM_COMPONENTS))\n",
    "\n",
    "    # iterate over all positions\n",
    "    for i in range(L):\n",
    "        # iterate over all substitutions\n",
    "        for A_i in range(num_symbols):\n",
    "            # iterate over couplings to all other sites\n",
    "            delta_hi = h_i[i, A_i] - h_i[i, target_seq[i]]\n",
    "            delta_Jij = 0.0\n",
    "\n",
    "            for j in range(L):\n",
    "                if i != j:\n",
    "                    delta_Jij += (\n",
    "                        J_ij[i, j, A_i, target_seq[j]] -\n",
    "                        J_ij[i, j, target_seq[i], target_seq[j]]\n",
    "                    )\n",
    "\n",
    "            H[i, A_i] = [delta_Jij + delta_hi, delta_Jij, delta_hi]\n",
    "\n",
    "    return H\n",
    "\n",
    "\n",
    "@jit(nopython=True)\n",
    "def _delta_hamiltonian(pos, subs, target_seq, J_ij, h_i):\n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    pos : np.array(int)\n",
    "        Vector of substituted positions\n",
    "    subs : np.array(int)\n",
    "        Vector of symbols above positions are substituted to\n",
    "    target_seq : np.array(int)\n",
    "        Target sequence for which mutant energy differences will be calculated\n",
    "        relative to\n",
    "    J_ij: np.array\n",
    "        L x L x num_symbols x num_symbols J_ij pair coupling parameter matrix\n",
    "    h_i: np.array\n",
    "        L x num_symbols h_i fields parameter matrix\n",
    "    Returns\n",
    "    -------\n",
    "    np.array\n",
    "        Vector of length 3, where elements correspond to delta of\n",
    "        1) total Hamiltonian and the 2) J_ij and 3) h_i sub-sums\n",
    "    \"\"\"\n",
    "    L, num_symbols = h_i.shape\n",
    "\n",
    "    M = pos.shape[0]\n",
    "    delta_hi = 0.0\n",
    "    delta_Jij = 0.0\n",
    "\n",
    "    # iterate over all changed positions\n",
    "    for m in range(M):\n",
    "        i = pos[m]\n",
    "        A_i = subs[m]\n",
    "\n",
    "        # change in fields\n",
    "        delta_hi += h_i[i, A_i] - h_i[i, target_seq[i]]\n",
    "\n",
    "        # couplings to all other positions in sequence\n",
    "        for j in range(L):\n",
    "            if i != j:\n",
    "                delta_Jij += (\n",
    "                    J_ij[i, j, A_i, target_seq[j]] -\n",
    "                    J_ij[i, j, target_seq[i], target_seq[j]]\n",
    "                )\n",
    "\n",
    "        # correct couplings between substituted positions:\n",
    "        # 1) do not count coupling twice (remove forward\n",
    "        #     and backward coupling)\n",
    "        # 2) adjust background to new sequence\n",
    "        for n in range(m + 1, M):\n",
    "            j = pos[n]\n",
    "            A_j = subs[n]\n",
    "            # remove forward and backward coupling delta\n",
    "            delta_Jij -= J_ij[i, j, A_i, target_seq[j]]\n",
    "            delta_Jij -= J_ij[i, j, target_seq[i], A_j]\n",
    "            delta_Jij += J_ij[i, j, target_seq[i], target_seq[j]]\n",
    "            # the following line cancels out with line further down:\n",
    "            # delta_Jij += J_ij[i, j, target_seq[i], target_seq[j]]\n",
    "\n",
    "            # now add coupling delta once in correct background\n",
    "            delta_Jij += J_ij[i, j, A_i, A_j]\n",
    "            # following line cancels out with line above:\n",
    "            # delta_Jij -= J_ij[i, j, target_seq[i], target_seq[j]]\n",
    "\n",
    "    return np.array([delta_Jij + delta_hi, delta_Jij, delta_hi])\n",
    "\n",
    "\n",
    "@jit(nopython=True)\n",
    "def _zero_sum_gauge(J_ij, inplace=False):\n",
    "    \"\"\"\n",
    "    Transform coupling matrix into zero-sum gauge\n",
    "    (i.e., row and column sums of each ij submatrix are 0)\n",
    "    Parameters\n",
    "    ----------\n",
    "    J_ij : np.array\n",
    "        Coupling matrix of size L x L x num_symbols x num_symbols\n",
    "        that should be transformed into zero-sum gauge\n",
    "    inplace : bool, optional (default: False)\n",
    "        Modify original matrix (True), or return transformed\n",
    "        matrix in a new matrix\n",
    "    Returns\n",
    "    -------\n",
    "    J_ij_0 : np.array\n",
    "        J_ij transformed into zero-sum gauge\n",
    "    \"\"\"\n",
    "    L, L2, num_symbols, num_symbols2 = J_ij.shape\n",
    "    assert L == L2 and num_symbols == num_symbols2\n",
    "\n",
    "    if inplace:\n",
    "        J_ij_0 = J_ij\n",
    "    else:\n",
    "        J_ij_0 = np.zeros((L, L, num_symbols, num_symbols))\n",
    "\n",
    "    # go through all pairs of positions\n",
    "    for i in range(L - 1):\n",
    "        for j in range(i + 1, L):\n",
    "            ij_mat = J_ij[i, j]\n",
    "\n",
    "            # calculate matrix, row and column averages\n",
    "            avg_ab = np.mean(ij_mat)\n",
    "\n",
    "            # can't use axis argument of np.mean in numba,\n",
    "            # so have to calculate rows/cols manually\n",
    "            avg_a = np.zeros(num_symbols)\n",
    "            avg_b = np.zeros(num_symbols)\n",
    "            ij_mat_T = ij_mat.T\n",
    "\n",
    "            for k in range(num_symbols):\n",
    "                avg_a[k] = np.mean(ij_mat[k])\n",
    "                avg_b[k] = np.mean(ij_mat_T[k])\n",
    "\n",
    "            # subtract correction terms from each entry\n",
    "            for a in range(num_symbols):\n",
    "                for b in range(num_symbols):\n",
    "                    J_ij_0[i, j, a, b] = (\n",
    "                        ij_mat[a, b] - avg_a[a] - avg_b[b] + avg_ab\n",
    "                    )\n",
    "                    J_ij_0[j, i, b, a] = J_ij_0[i, j, a, b]\n",
    "\n",
    "    return J_ij_0\n",
    "\n",
    "\n",
    "class CouplingsModel_1:\n",
    "    \"\"\"\n",
    "    Class to store parameters of pairwise undirected graphical model of sequences\n",
    "    and compute evolutionary couplings, sequence statistical energies, etc.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, filename, precision=\"float32\", file_format=\"plmc_v2\", **kwargs):\n",
    "        \"\"\"\n",
    "        Initializes the object with raw values read from binary .Jij file\n",
    "        Parameters\n",
    "        ----------\n",
    "        filename : str\n",
    "            Binary Jij file containing model parameters from plmc software\n",
    "        precision : {\"float32\", \"float64\"}, default: \"float32\"\n",
    "            Sets if input file has single (float32) or double precision (float64)\n",
    "        }\n",
    "        file_format : {\"plmc_v2\", \"plmc_v1\"}, default: \"plmc_v2\"\n",
    "            File format of parameter file.\n",
    "            Note: The use of \"plmc_v1\" is discouraged and only for backwards\n",
    "            compatibility as this format lacks crucial information about\n",
    "            parameters used by this class. Users are responsible for supplying\n",
    "            the missing values (e.g. regularization strength, alphabet or M_eff)\n",
    "            manually via the respective member variables/properties.\n",
    "        TODO: constructor for CouplingsModel (and all the other subclasses)\n",
    "        should be able to take a filehandle so one can read them from arbitrary streams\n",
    "        \"\"\"\n",
    "        if file_format == \"plmc_v2\":\n",
    "            self.__read_plmc_v2(filename, precision)\n",
    "        elif file_format == \"plmc_v1\":\n",
    "            self.__read_plmc_v1(\n",
    "                filename, precision, kwargs.get(\"alphabet\", None)\n",
    "            )\n",
    "        else:\n",
    "            raise ValueError(\n",
    "                \"Illegal file format {}, valid options are:\"\n",
    "                \"plmc_v2, plmc_v1\".format(\n",
    "                    file_format\n",
    "                )\n",
    "            )\n",
    "\n",
    "        self.alphabet_map = {s: i for i, s in enumerate(self.alphabet)}\n",
    "\n",
    "        # in non-gap mode, focus sequence is still coded with a gap character,\n",
    "        # but gap is not part of model alphabet anymore; so if mapping crashes\n",
    "        # that means there is a non-alphabet character in sequence array\n",
    "        # and therefore there is no focus sequence.\n",
    "        try:\n",
    "            self.target_seq_mapped = np.array([self.alphabet_map[x] for x in self.target_seq])\n",
    "            self.has_target_seq = (np.sum(self.target_seq_mapped) > 0)\n",
    "        except KeyError:\n",
    "            self.target_seq_mapped = np.zeros((self.L), dtype=np.int32)\n",
    "            self.has_target_seq = False\n",
    "\n",
    "        self._reset_precomputed()\n",
    "\n",
    "    def _reset_precomputed(self):\n",
    "        \"\"\"\n",
    "        Delete precomputed values (e.g. mutation matrices)\n",
    "        \"\"\"\n",
    "        self._single_mut_mat_full = None\n",
    "        self._double_mut_mat = None\n",
    "        self._cn_scores = None\n",
    "        self._fn_scores = None\n",
    "        self._mi_scores_raw = None\n",
    "        self._mi_scores_apc = None\n",
    "        self._ecs = None\n",
    "\n",
    "    def __read_plmc_v2(self, filename, precision):\n",
    "        \"\"\"\n",
    "        Read updated Jij file format from plmc.\n",
    "        Parameters\n",
    "        ----------\n",
    "        filename : str\n",
    "            Binary Jij file containing model parameters\n",
    "        precision : {\"float32\", \"float64\"}\n",
    "            Sets if input file has single or double precision\n",
    "        \"\"\"\n",
    "        with open(filename, \"rb\") as f:\n",
    "            # model length, number of symbols, valid/invalid sequences\n",
    "            # and iterations\n",
    "            self.L, self.num_symbols, self.N_valid, self.N_invalid, self.num_iter = (\n",
    "                np.fromfile(f, \"int32\", 5)\n",
    "            )\n",
    "\n",
    "            # theta, regularization weights, and effective number of samples\n",
    "            self.theta, self.lambda_h, self.lambda_J, self.lambda_group, self.N_eff = (\n",
    "                np.fromfile(f, precision, 5)\n",
    "            )\n",
    "\n",
    "            # Read alphabet (make sure we get proper unicode rather than byte string)\n",
    "            self.alphabet = np.fromfile(\n",
    "                f, \"S1\", self.num_symbols\n",
    "            ).astype(\"U1\")\n",
    "\n",
    "            # weights of individual sequences (after clustering)\n",
    "            self.weights = np.fromfile(\n",
    "                f, precision, self.N_valid + self.N_invalid\n",
    "            )\n",
    "\n",
    "            # target sequence and index mapping, again ensure unicode\n",
    "            self._target_seq = np.fromfile(f, \"S1\", self.L).astype(\"U1\")\n",
    "            self.index_list = np.fromfile(f, \"int32\", self.L)\n",
    "\n",
    "            # single site frequencies f_i and fields h_i\n",
    "            self.f_i, = np.fromfile(\n",
    "                f, dtype=(precision, (self.L, self.num_symbols)), count=1\n",
    "            )\n",
    "\n",
    "            self.h_i, = np.fromfile(\n",
    "                f, dtype=(precision, (self.L, self.num_symbols)), count=1\n",
    "            )\n",
    "\n",
    "            # pair frequencies f_ij and pair couplings J_ij / J_ij\n",
    "            self.f_ij = np.zeros(\n",
    "                (self.L, self.L, self.num_symbols, self.num_symbols)\n",
    "            )\n",
    "\n",
    "            self.J_ij = np.zeros(\n",
    "                (self.L, self.L, self.num_symbols, self.num_symbols)\n",
    "            )\n",
    "\n",
    "            # TODO: could read triangle matrix from file in one block\n",
    "            # like in read_params.m, which would result in faster reading\n",
    "            # but also 50% higher memory usage... for now save memory\n",
    "            for i in range(self.L - 1):\n",
    "                for j in range(i + 1, self.L):\n",
    "#                     self.f_ij[i, j], = np.fromfile(\n",
    "#                         f, dtype=(precision, (self.num_symbols, self.num_symbols)),\n",
    "#                         count=1\n",
    "#                     )\n",
    "                    self.f_ij[j, i] = self.f_ij[i, j].T\n",
    "\n",
    "            for i in range(self.L - 1):\n",
    "                for j in range(i + 1, self.L):\n",
    "                    self.J_ij[i, j], = np.fromfile(\n",
    "                        f, dtype=(precision, (self.num_symbols, self.num_symbols)),\n",
    "                        count=1\n",
    "                    )\n",
    "                    self.J_ij[j, i] = self.J_ij[i, j].T\n",
    "\n",
    "            # if lambda_h is negative, the model was\n",
    "            # inferred using mean-field\n",
    "            if self.lambda_h < 0:\n",
    "                # cast model to mean field model\n",
    "                from evcouplings.couplings.mean_field import MeanFieldCouplingsModel\n",
    "                self.__class__ = MeanFieldCouplingsModel\n",
    "\n",
    "                # handle requirements specific to\n",
    "                # the mean-field couplings model\n",
    "                self.transform_from_plmc_model()\n",
    "\n",
    "    def __read_plmc_v1(self, filename, precision, alphabet=None):\n",
    "        \"\"\"\n",
    "        Read original eij/Jij file format from plmc. Use of this old format\n",
    "        is discouraged (see constructor documentation for details)\n",
    "        Parameters\n",
    "        ----------\n",
    "        filename : str\n",
    "            Binary Jij file containing model parameters\n",
    "        precision : {\"float32\", \"float64\"}\n",
    "            Sets if input file has single or double precision\n",
    "        alphabet : str, default: None\n",
    "        \"\"\"\n",
    "        # local definitions of alphabets for backwards compatibility\n",
    "        # better: use v2 file format which includes this information\n",
    "        GAP = \"-\"\n",
    "        ALPHABET_PROTEIN_NOGAP = \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "        ALPHABET_PROTEIN = GAP + ALPHABET_PROTEIN_NOGAP\n",
    "\n",
    "        with open(filename, \"rb\") as f:\n",
    "            # model length, number of symbols\n",
    "            self.L, = np.fromfile(f, \"int32\", 1)\n",
    "            self.num_symbols, = np.fromfile(f, \"int32\", 1)\n",
    "\n",
    "            # Old format does not have alphabet in file, so need\n",
    "            # to guess it or use user-supplied alphabet.\n",
    "            # if no alphabet is given, try to guess\n",
    "            if alphabet is None:\n",
    "                if self.num_symbols == 21:\n",
    "                    alphabet = ALPHABET_PROTEIN\n",
    "                elif self.num_symbols == 20:\n",
    "                    alphabet = ALPHABET_PROTEIN_NOGAP\n",
    "                else:\n",
    "                    raise ValueError(\n",
    "                        \"Could not guess default alphabet for \"\n",
    "                        \"{} states, specify alphabet parameter.\".format(\n",
    "                            self.num_symbols\n",
    "                        )\n",
    "                    )\n",
    "            else:\n",
    "                # verify if size of given alphabet matches model\n",
    "                if len(alphabet) != self.num_symbols:\n",
    "                    raise ValueError(\n",
    "                        \"Size of alphabet ({}) does not agree with \"\n",
    "                        \"number of states in model ({})\".format(\n",
    "                            len(alphabet), self.num_symbols\n",
    "                        )\n",
    "                    )\n",
    "\n",
    "            self.alphabet = np.array(list(alphabet))\n",
    "\n",
    "            # target sequence and index mapping, again ensure unicode\n",
    "            self._target_seq = np.fromfile(f, \"S1\", self.L).astype(\"U1\")\n",
    "            self.index_list = np.fromfile(f, \"int32\", self.L)\n",
    "\n",
    "            # set all the information missing in v1 files to None\n",
    "\n",
    "            # valid/invalid sequences, number of iterations\n",
    "            self.N_valid = None\n",
    "            self.N_invalid = None\n",
    "            self.num_iter = None\n",
    "\n",
    "            # theta, regularization weights, and effective number of samples\n",
    "            self.theta = None\n",
    "            self.lambda_h = None\n",
    "            self.lambda_J = None\n",
    "            self.lambda_group = None\n",
    "            self.N_eff = None\n",
    "\n",
    "            # weights of individual sequences (after clustering)\n",
    "            self.weights = None\n",
    "\n",
    "            # single site frequencies f_i and fields h_i\n",
    "            self.f_i, = np.fromfile(\n",
    "                f, dtype=(precision, (self.L, self.num_symbols)), count=1\n",
    "            )\n",
    "\n",
    "            self.h_i, = np.fromfile(\n",
    "                f, dtype=(precision, (self.L, self.num_symbols)), count=1\n",
    "            )\n",
    "\n",
    "            # pair frequencies f_ij and pair couplings J_ij / J_ij\n",
    "            self.f_ij = np.zeros(\n",
    "                (self.L, self.L, self.num_symbols, self.num_symbols)\n",
    "            )\n",
    "\n",
    "            self.J_ij = np.zeros(\n",
    "                (self.L, self.L, self.num_symbols, self.num_symbols)\n",
    "            )\n",
    "\n",
    "            for i in range(self.L - 1):\n",
    "                for j in range(i + 1, self.L):\n",
    "                    file_i, file_j = np.fromfile(f, \"int32\", 2)\n",
    "\n",
    "#                     if i + 1 != file_i or j + 1 != file_j:\n",
    "#                         raise ValueError(\n",
    "#                             \"Error: column pair indices inconsistent. \"\n",
    "#                             \"Expected: {} {}; File: {} {}\".format(i + 1, j + 1, file_i, file_j)\n",
    "#                         )\n",
    "\n",
    "                    self.f_ij[i, j], = np.fromfile(\n",
    "                        f, dtype=(precision, (self.num_symbols, self.num_symbols)),\n",
    "                        count=1\n",
    "                    )\n",
    "                    self.f_ij[j, i] = self.f_ij[i, j].T\n",
    "\n",
    "                    self.J_ij[i, j], = np.fromfile(\n",
    "                        f, dtype=(precision, (self.num_symbols, self.num_symbols)),\n",
    "                        count=1\n",
    "                    )\n",
    "                    self.J_ij[j, i] = self.J_ij[i, j].T\n",
    "\n",
    "    @property\n",
    "    def target_seq(self):\n",
    "        \"\"\"\n",
    "        Target/Focus sequence of model used for delta_hamiltonian\n",
    "        calculations (including single and double mutation matrices)\n",
    "        \"\"\"\n",
    "        return self._target_seq\n",
    "\n",
    "    @target_seq.setter\n",
    "    def target_seq(self, sequence):\n",
    "        \"\"\"\n",
    "        Define a new target sequence\n",
    "        Parameters\n",
    "        ----------\n",
    "        sequence : str, or list of chars\n",
    "            Define a new default sequence for relative Hamiltonian\n",
    "            calculations (e.g. energy difference relative to wild-type\n",
    "            sequence). Length of sequence must correspond to model length (self.L)\n",
    "        \"\"\"\n",
    "        self._reset_precomputed()\n",
    "\n",
    "        if len(sequence) != self.L:\n",
    "            raise ValueError(\n",
    "                \"Sequence length inconsistent with model length: {} {}\".format(\n",
    "                    len(sequence), self.L\n",
    "                )\n",
    "            )\n",
    "\n",
    "        if isinstance(sequence, str):\n",
    "            sequence = list(sequence)\n",
    "\n",
    "        self._target_seq = np.array(sequence)\n",
    "        self.target_seq_mapped = np.array([self.alphabet_map[x] for x in self.target_seq])\n",
    "        self.has_target_seq = True\n",
    "\n",
    "    @property\n",
    "    def index_list(self):\n",
    "        \"\"\"\n",
    "        Target/Focus sequence of model used for delta_hamiltonian\n",
    "        calculations (including single and double mutation matrices)\n",
    "        \"\"\"\n",
    "        return self._index_list\n",
    "\n",
    "    @index_list.setter\n",
    "    def index_list(self, mapping):\n",
    "        \"\"\"\n",
    "        Define a new number mapping for sequences\n",
    "        Parameters\n",
    "        ----------\n",
    "        mapping: list of int\n",
    "            Sequence indices of the positions in the model.\n",
    "            Length of list must correspond to model length (self.L)\n",
    "        \"\"\"\n",
    "        if len(mapping) != self.L:\n",
    "            raise ValueError(\n",
    "                \"Mapping length inconsistent with model length: {} {}\".format(\n",
    "                    len(mapping), self.L\n",
    "                )\n",
    "            )\n",
    "\n",
    "        self._index_list = deepcopy(mapping)\n",
    "        self.index_map = {b: a for a, b in enumerate(self.index_list)}\n",
    "\n",
    "        # update ECs, if they were already calculated\n",
    "        if hasattr(self, \"_ecs\"):\n",
    "            self._calculate_ecs()\n",
    "\n",
    "    def convert_sequences(self, sequences):\n",
    "        \"\"\"\n",
    "        Converts sequences in string format into internal symbol representation\n",
    "        according to alphabet of model\n",
    "        Parameters\n",
    "        ----------\n",
    "        sequences : list of str\n",
    "            List of sequences (must have same length and correspond to\n",
    "            model states)\n",
    "        Returns\n",
    "        -------\n",
    "        np.array\n",
    "            Matrix of size len(sequences) x L of sequences converted to\n",
    "            integer symbols\n",
    "        \"\"\"\n",
    "        seq_lens = list(set(map(len, sequences)))\n",
    "        if len(seq_lens) != 1:\n",
    "            raise ValueError(\"Input sequences have different lengths: \" + str(seq_lens))\n",
    "\n",
    "        L_seq = seq_lens[0]\n",
    "        if L_seq != self.L:\n",
    "            raise ValueError(\n",
    "                \"Sequence lengths do not correspond to model length: {} {}\".format(\n",
    "                    L_seq, self.L\n",
    "                )\n",
    "            )\n",
    "\n",
    "        S = np.empty((len(sequences), L_seq), dtype=np.int)\n",
    "\n",
    "        try:\n",
    "            for i, s in enumerate(sequences):\n",
    "                S[i] = [self.alphabet_map[x] for x in s]\n",
    "        except KeyError:\n",
    "            raise ValueError(\n",
    "                \"Invalid symbol in sequence {}: {}\".format(i, s)\n",
    "            )\n",
    "\n",
    "        return S\n",
    "\n",
    "    def hamiltonians(self, sequences):\n",
    "        \"\"\"\n",
    "        Calculates the Hamiltonians of the global probability distribution P(A_1, ..., A_L)\n",
    "        for the given sequences A_1,...,A_L from J_ij and h_i parameters\n",
    "        Parameters\n",
    "        ----------\n",
    "        sequences : list of str\n",
    "            List of sequences for which Hamiltonian will be computed,\n",
    "            or converted np.array obtained using convert_sequences method\n",
    "        Returns\n",
    "        -------\n",
    "        np.array\n",
    "            Float matrix of size len(sequences) x 3, where each row corresponds to the\n",
    "            1) total Hamiltonian of sequence and the 2) J_ij and 3) h_i sub-sums\n",
    "        \"\"\"\n",
    "        if isinstance(sequences, list):\n",
    "            sequences = self.convert_sequences(sequences)\n",
    "\n",
    "        return _hamiltonians(sequences, self.J_ij, self.h_i)\n",
    "\n",
    "    @property\n",
    "    def single_mut_mat_full(self):\n",
    "        \"\"\"\n",
    "        Hamiltonian difference for all possible single-site variants\n",
    "        L x num_symbol x 3 matrix (np.array) containing delta Hamiltonians\n",
    "        for all possible single mutants of target sequence.\n",
    "        Third dimension: 1) full Hamiltonian, 2) J_ij, 3) h_i\n",
    "        \"\"\"\n",
    "        if self._single_mut_mat_full is None:\n",
    "            self._single_mut_mat_full = _single_mutant_hamiltonians(\n",
    "                self.target_seq_mapped, self.J_ij, self.h_i\n",
    "            )\n",
    "\n",
    "        return self._single_mut_mat_full\n",
    "\n",
    "    @property\n",
    "    def single_mut_mat(self):\n",
    "        \"\"\"\n",
    "        Hamiltonian difference for all possible single-site variants\n",
    "        L x num_symbol matrix (np.array) containing delta Hamiltonians\n",
    "        for all possible single mutants of target sequence.\n",
    "        \"\"\"\n",
    "        return self.single_mut_mat_full[:, :, FULL]\n",
    "\n",
    "    def delta_hamiltonian(self, substitutions, verify_mutants=True):\n",
    "        \"\"\"\n",
    "        Calculate difference in statistical energy relative to\n",
    "        self.target_seq by changing sequence according to list of\n",
    "        substitutions\n",
    "        Parameters\n",
    "        ----------\n",
    "        substitutions : list of tuple(pos, subs_from, subs_to)\n",
    "            Substitutions to be applied to target sequence\n",
    "        verify_mutants : bool, optional\n",
    "            Test if subs_from is consistent with self.target_seq\n",
    "        Returns\n",
    "        -------\n",
    "        np.array\n",
    "            Vector of length 3 with 1) total delta Hamiltonian,\n",
    "            2) delta J_ij, 3) delta h_i\n",
    "        \"\"\"\n",
    "        pos = np.empty(len(substitutions), dtype=np.int)\n",
    "        subs = np.empty(len(substitutions), dtype=np.int)\n",
    "\n",
    "        try:\n",
    "            for i, (subs_pos, subs_from, subs_to) in enumerate(substitutions):\n",
    "                pos[i] = self.index_map[subs_pos]\n",
    "                subs[i] = self.alphabet_map[subs_to]\n",
    "                if verify_mutants and subs_from != self.target_seq[pos[i]]:\n",
    "                    raise ValueError(\n",
    "                        \"Inconsistency with target sequence: pos={} target={} subs={}\".format(\n",
    "                            subs_pos, self.target_seq[i], subs_from\n",
    "                        )\n",
    "                    )\n",
    "        except KeyError:\n",
    "            raise ValueError(\n",
    "                \"Illegal substitution: {}{}{}\\nAlphabet: {}\\nPositions: {}\".format(\n",
    "                    subs_from, subs_pos, subs_to, self.alphabet_map, self.index_list\n",
    "                )\n",
    "            )\n",
    "\n",
    "        return _delta_hamiltonian(pos, subs, self.target_seq_mapped, self.J_ij, self.h_i)\n",
    "\n",
    "    @property\n",
    "    def double_mut_mat(self):\n",
    "        \"\"\"\n",
    "        Hamiltonian difference for all possible double mutant variants\n",
    "        L x L x num_symbol x num_symbol matrix containing delta Hamiltonians\n",
    "        for all possible double mutants of target sequence\n",
    "        \"\"\"\n",
    "        if self._double_mut_mat is None:\n",
    "            self._double_mut_mat = np.zeros(\n",
    "                (self.L, self.L, self.num_symbols, self.num_symbols)\n",
    "            )\n",
    "\n",
    "            seq = self.target_seq_mapped\n",
    "            for i in range(self.L - 1):\n",
    "                for j in range(i + 1, self.L):\n",
    "                    self._double_mut_mat[i, j] = (\n",
    "                        np.tile(self.single_mut_mat[i], (self.num_symbols, 1)).T +\n",
    "                        np.tile(self.single_mut_mat[j], (self.num_symbols, 1)) +\n",
    "                        self.J_ij[i, j] -\n",
    "                        np.tile(self.J_ij[i, j, :, seq[j]], (self.num_symbols, 1)).T -\n",
    "                        np.tile(self.J_ij[i, j, seq[i], :], (self.num_symbols, 1)) +\n",
    "                        # we are only interested in difference to WT, so normalize\n",
    "                        # for second couplings subtraction with last term\n",
    "                        self.J_ij[i, j, seq[i], seq[j]])\n",
    "\n",
    "                    self._double_mut_mat[j, i] = self._double_mut_mat[i, j].T\n",
    "\n",
    "        return self._double_mut_mat\n",
    "\n",
    "    @classmethod\n",
    "    def apc(cls, matrix):\n",
    "        \"\"\"\n",
    "        Apply average product correction (Dunn et al., Bioinformatics, 2008)\n",
    "        to matrix\n",
    "        Parameters\n",
    "        ----------\n",
    "        matrix : np.array\n",
    "            Symmetric L x L matrix which should be corrected by APC\n",
    "        Returns\n",
    "        -------\n",
    "        np.array\n",
    "            Symmetric L x L matrix with APC correction applied\n",
    "        \"\"\"\n",
    "        L = matrix.shape[0]\n",
    "        if L != matrix.shape[1]:\n",
    "            raise ValueError(\"Input matrix is not symmetric: {}\".format(matrix.shape))\n",
    "\n",
    "        col_means = np.mean(matrix, axis=0) * L / (L - 1)\n",
    "        matrix_mean = np.mean(matrix) * L / (L - 1)\n",
    "\n",
    "        apc = np.dot(\n",
    "            col_means.reshape(L, 1), col_means.reshape(1, L)\n",
    "        ) / matrix_mean\n",
    "\n",
    "        # subtract APC and blank diagonal entries\n",
    "        corrected_matrix = matrix - apc\n",
    "        corrected_matrix[np.diag_indices(L)] = 0\n",
    "\n",
    "        return corrected_matrix\n",
    "\n",
    "    def _calculate_ecs(self):\n",
    "        \"\"\"\n",
    "        Calculates FN and CN scores as defined in Ekeberg et al., Phys Rev E, 2013,\n",
    "        as well as MI scores.\n",
    "        \"\"\"\n",
    "        # calculate Frobenius norm for each pair of sites (i, j)\n",
    "        # also calculate mutual information\n",
    "        self._fn_scores = np.zeros((self.L, self.L))\n",
    "        self._mi_scores_raw = np.zeros((self.L, self.L))\n",
    "\n",
    "        # transform couplings into zero-sum gauge\n",
    "        J_ij_0 = _zero_sum_gauge(self.J_ij)\n",
    "\n",
    "        for i in range(self.L - 1):\n",
    "            for j in range(i + 1, self.L):\n",
    "                self._fn_scores[i, j] = np.linalg.norm(J_ij_0[i, j], \"fro\")\n",
    "                self._fn_scores[j, i] = self._fn_scores[i, j]\n",
    "\n",
    "                # mutual information\n",
    "                p = self.f_ij[i, j]\n",
    "                m = np.dot(self.f_i[i, np.newaxis].T, self.f_i[j, np.newaxis])\n",
    "                self._mi_scores_raw[i, j] = np.sum(p[p > 0] * np.log(p[p > 0] / m[p > 0]))\n",
    "                self._mi_scores_raw[j, i] = self._mi_scores_raw[i, j]\n",
    "\n",
    "        # apply Average Product Correction (Dunn et al., Bioinformatics, 2008)\n",
    "        # subtract APC and blank diagonal entries\n",
    "        self._cn_scores = self.apc(self._fn_scores)\n",
    "        self._mi_scores_apc = self.apc(self._mi_scores_raw)\n",
    "\n",
    "        # create internal dataframe representation\n",
    "        ecs = []\n",
    "        for i in range(self.L - 1):\n",
    "            for j in range(i + 1, self.L):\n",
    "                # if we have custom indeces, cannot compute sequence distance\n",
    "                # easily, unless we use segment information\n",
    "                try:\n",
    "                    seqdist = abs(self.index_list[i] - self.index_list[j])\n",
    "                except TypeError:\n",
    "                    seqdist = np.nan\n",
    "\n",
    "                ecs.append((\n",
    "                    self.index_list[i], self.target_seq[i],\n",
    "                    self.index_list[j], self.target_seq[j],\n",
    "                    seqdist,\n",
    "                    self._mi_scores_raw[i, j], self._mi_scores_apc[i, j],\n",
    "                    self._fn_scores[i, j], self._cn_scores[i, j]\n",
    "                ))\n",
    "\n",
    "        self._ecs = pd.DataFrame(\n",
    "            ecs, columns=[\"i\", \"A_i\", \"j\", \"A_j\", \"seqdist\", \"mi_raw\", \"mi_apc\", \"fn\", \"cn\"]\n",
    "        ).sort_values(by=\"cn\", ascending=False)\n",
    "\n",
    "    @property\n",
    "    def cn_scores(self):\n",
    "        \"\"\"\n",
    "        L x L numpy matrix with CN (corrected norm) scores\n",
    "        \"\"\"\n",
    "        if self._cn_scores is None:\n",
    "            self._calculate_ecs()\n",
    "\n",
    "        return self._cn_scores\n",
    "\n",
    "    @property\n",
    "    def fn_scores(self):\n",
    "        \"\"\"\n",
    "        L x L numpy matrix with FN (Frobenius norm) scores\n",
    "        \"\"\"\n",
    "        if self._fn_scores is None:\n",
    "            self._calculate_ecs()\n",
    "\n",
    "        return self._fn_scores\n",
    "\n",
    "    @property\n",
    "    def mi_scores_raw(self):\n",
    "        \"\"\"\n",
    "        L x L numpy matrix with MI (mutual information) scores\n",
    "        without APC correction\n",
    "        \"\"\"\n",
    "        if self._mi_scores_raw is None:\n",
    "            self._calculate_ecs()\n",
    "\n",
    "        return self._mi_scores_raw\n",
    "\n",
    "    @property\n",
    "    def mi_scores_apc(self):\n",
    "        \"\"\"\n",
    "        L x L numpy matrix with MI (mutual information) scores\n",
    "        with APC correction\n",
    "        \"\"\"\n",
    "        if self._mi_scores_apc is None:\n",
    "            self._calculate_ecs()\n",
    "\n",
    "        return self._mi_scores_apc\n",
    "\n",
    "    @property\n",
    "    def ecs(self):\n",
    "        \"\"\"\n",
    "        DataFrame with evolutionary couplings, sorted by CN score\n",
    "        (all scores: CN, FN, MI)\n",
    "        \"\"\"\n",
    "        if self._ecs is None:\n",
    "            self._calculate_ecs()\n",
    "\n",
    "        return self._ecs\n",
    "\n",
    "    def to_independent_model(self):\n",
    "        \"\"\"\n",
    "        Estimate parameters of a single-site model using\n",
    "        Gaussian prior/L2 regularization.\n",
    "        Returns\n",
    "        -------\n",
    "        CouplingsModel\n",
    "            Copy of object turned into independent model\n",
    "        \"\"\"\n",
    "        from scipy.optimize import fmin_bfgs\n",
    "\n",
    "        def _log_post(x, *args):\n",
    "            \"\"\"\n",
    "            Log posterior of single-site model\n",
    "            \"\"\"\n",
    "            (fi, lambda_h, N) = args\n",
    "            logZ = np.log(np.exp(x).sum())\n",
    "            return N * (logZ - (fi * x).sum()) + lambda_h * ((x ** 2).sum())\n",
    "\n",
    "        def _gradient(x, *args):\n",
    "            \"\"\"\n",
    "            Gradient of single-site model\n",
    "            \"\"\"\n",
    "            (fi, lambda_h, N) = args\n",
    "            Z = np.exp(x).sum()\n",
    "            P = np.exp(x) / Z\n",
    "            return N * (P - fi) + lambda_h * 2 * x\n",
    "\n",
    "        h_i = np.zeros((self.L, self.num_symbols))\n",
    "\n",
    "        for i in range(self.L):\n",
    "            x0 = np.zeros(self.num_symbols)\n",
    "            h_i[i] = fmin_bfgs(\n",
    "                _log_post, x0, _gradient,\n",
    "                args=(self.f_i[i], self.lambda_h, self.N_eff),\n",
    "                disp=False\n",
    "            )\n",
    "\n",
    "        c0 = deepcopy(self)\n",
    "        c0.h_i = h_i\n",
    "        c0.J_ij.fill(0)\n",
    "        c0._reset_precomputed()\n",
    "        return c0\n",
    "\n",
    "    # syntactic sugar to access most important member variables in target numbering space\n",
    "\n",
    "    def __map(self, indices, mapping):\n",
    "        \"\"\"\n",
    "        Applies a mapping either to a single index, or to a list of indices\n",
    "        Parameters\n",
    "        ----------\n",
    "        indices : Iterable, or single item of any type\n",
    "            Items that should be mapped into new space.\n",
    "            Tuples and strings will not be used as an\n",
    "            Iterable, but as single items.\n",
    "        mapping : dict\n",
    "            Mapping into new space\n",
    "        Returns\n",
    "        -------\n",
    "        Iterable, or single item\n",
    "            Items mapped into new space\n",
    "        \"\"\"\n",
    "        is_sequence = (\n",
    "            isinstance(indices, Iterable) and\n",
    "            not isinstance(indices, str) and\n",
    "            not isinstance(indices, tuple)\n",
    "        )\n",
    "\n",
    "        if is_sequence:\n",
    "            return np.array(\n",
    "                [mapping[i] for i in indices]\n",
    "            )\n",
    "        else:\n",
    "            return mapping[indices]\n",
    "\n",
    "    def __4d_access(self, matrix, i=None, j=None, A_i=None, A_j=None):\n",
    "        \"\"\"\n",
    "        Provides shortcut access to column pair properties\n",
    "        (e.g. J_ij or f_ij matrices)\n",
    "        Parameters\n",
    "        -----------\n",
    "        i : Iterable(int) or int\n",
    "            Position(s) on first matrix axis\n",
    "        j : Iterable(int) or int\n",
    "            Position(s) on second matrix axis\n",
    "        A_i : Iterable(str) or str\n",
    "            Symbols corresponding to first matrix axis\n",
    "        A_j : Iterable(str) or str\n",
    "            Symbols corresponding to second matrix axis\n",
    "        Returns\n",
    "        -------\n",
    "        np.array\n",
    "            4D matrix \"matrix\" sliced according to values i, j, A_i and A_j\n",
    "        \"\"\"\n",
    "        i = self.__map(i, self.index_map) if i is not None else _SLICE\n",
    "        j = self.__map(j, self.index_map) if j is not None else _SLICE\n",
    "        A_i = self.__map(A_i, self.alphabet_map) if A_i is not None else _SLICE\n",
    "        A_j = self.__map(A_j, self.alphabet_map) if A_j is not None else _SLICE\n",
    "        return matrix[i, j, A_i, A_j]\n",
    "\n",
    "    def __2d_access(self, matrix, i=None, A_i=None):\n",
    "        \"\"\"\n",
    "        Provides shortcut access to single-column properties\n",
    "        (e.g. f_i or h_i matrices)\n",
    "        Parameters\n",
    "        -----------\n",
    "        i : Iterable(int) or int\n",
    "            Position(s) on first matrix axis\n",
    "        A_i : Iterable(str) or str\n",
    "            Symbols corresponding to first matrix axis\n",
    "        Returns\n",
    "        -------\n",
    "        np.array\n",
    "            2D matrix \"matrix\" sliced according to values i and A_i\n",
    "        \"\"\"\n",
    "        i = self.__map(i, self.index_map) if i is not None else _SLICE\n",
    "        A_i = self.__map(A_i, self.alphabet_map) if A_i is not None else _SLICE\n",
    "        return matrix[i, A_i]\n",
    "\n",
    "    def __2d_access_score_matrix(self, matrix, i=None, j=None):\n",
    "        \"\"\"\n",
    "        Provides shortcut access to quadratic 2D matrices\n",
    "        Parameters\n",
    "        -----------\n",
    "        i : Iterable(int) or int\n",
    "            Position(s) on first matrix axis\n",
    "        j : Iterable(int) or int\n",
    "            Position(s) on first matrix axis\n",
    "        Returns\n",
    "        -------\n",
    "        np.array\n",
    "            2D matrix \"matrix\" sliced according to values i and j\n",
    "        \"\"\"\n",
    "        i = self.__map(i, self.index_map) if i is not None else _SLICE\n",
    "        j = self.__map(j, self.index_map) if j is not None else _SLICE\n",
    "        return matrix[i, j]\n",
    "\n",
    "    def Jij(self, i=None, j=None, A_i=None, A_j=None):\n",
    "        \"\"\"\n",
    "        Quick access to J_ij matrix with automatic index mapping.\n",
    "        See __4d_access for explanation of parameters.\n",
    "        \"\"\"\n",
    "        return self.__4d_access(self.J_ij, i, j, A_i, A_j)\n",
    "\n",
    "    def fij(self, i=None, j=None, A_i=None, A_j=None):\n",
    "        \"\"\"\n",
    "        Quick access to f_ij matrix with automatic index mapping.\n",
    "        See __4d_access for explanation of parameters.\n",
    "        \"\"\"\n",
    "        return self.__4d_access(self.f_ij, i, j, A_i, A_j)\n",
    "\n",
    "    def hi(self, i=None, A_i=None):\n",
    "        \"\"\"\n",
    "        Quick access to h_i matrix with automatic index mapping.\n",
    "        See __2d_access for explanation of parameters.\n",
    "        \"\"\"\n",
    "        return self.__2d_access(self.h_i, i, A_i)\n",
    "\n",
    "    def fi(self, i=None, A_i=None):\n",
    "        \"\"\"\n",
    "        Quick access to f_i matrix with automatic index mapping.\n",
    "        See __2d_access for explanation of parameters.\n",
    "        \"\"\"\n",
    "        return self.__2d_access(self.f_i, i, A_i)\n",
    "\n",
    "    def cn(self, i=None, j=None):\n",
    "        \"\"\"\n",
    "        Quick access to cn_scores matrix with automatic index mapping.\n",
    "        See __2d_access_score_matrix for explanation of parameters.\n",
    "        \"\"\"\n",
    "        return self.__2d_access_score_matrix(self.cn_scores, i, j)\n",
    "\n",
    "    def fn(self, i=None, j=None):\n",
    "        \"\"\"\n",
    "        Quick access to fn_scores matrix with automatic index mapping.\n",
    "        See __2d_access_score_matrix for explanation of parameters.\n",
    "        \"\"\"\n",
    "        return self.__2d_access_score_matrix(self.fn_scores, i, j)\n",
    "\n",
    "    def mi_apc(self, i=None, j=None):\n",
    "        \"\"\"\n",
    "        Quick access to mi_scores_apc matrix with automatic index mapping.\n",
    "        See __2d_access_score_matrix for explanation of parameters.\n",
    "        \"\"\"\n",
    "        return self.__2d_access_score_matrix(self.mi_scores_apc, i, j)\n",
    "\n",
    "    def mi_raw(self, i=None, j=None):\n",
    "        \"\"\"\n",
    "        Quick access to mi_scores_raw matrix with automatic index mapping.\n",
    "        See __2d_access_score_matrix for explanation of parameters.\n",
    "        \"\"\"\n",
    "        return self.__2d_access_score_matrix(self.mi_scores_raw, i, j)\n",
    "\n",
    "    def mn(self, i=None):\n",
    "        \"\"\"\n",
    "        Map model numbering to internal numbering\n",
    "        Parameters\n",
    "        ----------\n",
    "        i : Iterable(int) or int\n",
    "            Position(s) to be mapped from model numbering space\n",
    "            into internal numbering space\n",
    "        Returns\n",
    "        -------\n",
    "        Iterable(int) or int\n",
    "            Remapped position(s)\n",
    "        \"\"\"\n",
    "        if i is None:\n",
    "            return np.array(sorted(self.index_map.values()))\n",
    "        else:\n",
    "            return self.__map(i, self.index_map)\n",
    "\n",
    "    def mui(self, i=None):\n",
    "        \"\"\"\n",
    "        Legacy method for backwards compatibility. See self.mn for explanation.\n",
    "        \"\"\"\n",
    "        return self.mn(i)\n",
    "\n",
    "    def sn(self, i=None):\n",
    "        \"\"\"\n",
    "        Map internal numbering to sequence numbering\n",
    "        Parameters\n",
    "        ----------\n",
    "        i : Iterable(int) or int\n",
    "            Position(s) to be mapped from internal numbering space\n",
    "            into sequence numbering space.\n",
    "        Returns\n",
    "        -------\n",
    "        Iterable(int) or int\n",
    "            Remapped position(s)\n",
    "        \"\"\"\n",
    "        if i is None:\n",
    "            return np.array(self.index_list)\n",
    "        else:\n",
    "            return self.__map(i, self.index_list)\n",
    "\n",
    "    def itu(self, i=None):\n",
    "        \"\"\"\n",
    "        Legacy method for backwards compatibility. See self.sn for explanation.\n",
    "        \"\"\"\n",
    "        return self.sn(i)\n",
    "\n",
    "    def seq(self, i=None):\n",
    "        \"\"\"\n",
    "        Access target sequence of model\n",
    "        Parameters\n",
    "        ----------\n",
    "        i : Iterable(int) or int\n",
    "            Position(s) for which symbol should be retrieved\n",
    "        Returns\n",
    "        -------\n",
    "        Iterable(char) or char\n",
    "            Sequence symbols\n",
    "        \"\"\"\n",
    "        if i is None:\n",
    "            return self.target_seq\n",
    "        else:\n",
    "            i = self.__map(i, self.index_map)\n",
    "            return self.__map(i, self.target_seq)\n",
    "\n",
    "    def smm(self, i=None, A_i=None):\n",
    "        \"\"\"\n",
    "        Access delta_Hamiltonian matrix of single mutants of target sequence\n",
    "        Parameters\n",
    "        ----------\n",
    "        i : Iterable(int) or int\n",
    "            Position(s) for which energy difference should be retrieved\n",
    "        A_i : Iterable(char) or char\n",
    "            Substitutions for which energy difference should be retrieved\n",
    "        Returns\n",
    "        -------\n",
    "        np.array(float)\n",
    "            2D matrix containing energy differences for slices along both\n",
    "            axes of single mutation matrix (first axis: position, second\n",
    "            axis: substitution).\n",
    "        \"\"\"\n",
    "        return self.__2d_access(self.single_mut_mat, i, A_i)\n",
    "\n",
    "    def dmm(self, i=None, j=None, A_i=None, A_j=None):\n",
    "        \"\"\"\n",
    "        Access delta_Hamiltonian matrix of double mutants of target sequence\n",
    "        Parameters\n",
    "        ----------\n",
    "        i : Iterable(int) or int\n",
    "            Position(s) of first substitution(s)\n",
    "        j : Iterable(int) or int\n",
    "            Position(s) of second substitution(s)\n",
    "        A_i : Iterable(char) or char\n",
    "            Substitution(s) to first position\n",
    "        A_j : Iterable(char) or char\n",
    "            Substitution(s) to second position\n",
    "        Returns\n",
    "        -------\n",
    "        np.array(float)\n",
    "            4D matrix containing energy differences for slices along both\n",
    "            axes of double mutation matrix (axes 1/2: position, axis 3/4:\n",
    "            substitutions).\n",
    "        \"\"\"\n",
    "        return self.__4d_access(self.double_mut_mat, i, j, A_i, A_j)\n",
    "\n",
    "    def to_file(self, out_file, precision=\"float32\", file_format=\"plmc_v2\"):\n",
    "        \"\"\"\n",
    "        Writes the potentially modified model again to binary file\n",
    "        Parameters\n",
    "        ----------\n",
    "        out_file: str\n",
    "            A string specifying the path to a file\n",
    "        precision: {\"float16\", \"float32\", \"float64\"}, optional (default: \"float32\")\n",
    "            Numerical NumPy data type specifying the precision\n",
    "            used to write numerical values to file\n",
    "        file_format : {\"plmc_v1\", \"plmc_v2\"}, optional (default: \"plmc_v2\")\n",
    "            Available file formats\n",
    "        \"\"\"\n",
    "        new = file_format.lower() == \"plmc_v2\"\n",
    "        with open(out_file, \"wb\") as f:\n",
    "            np.array([self.L, self.num_symbols], dtype=\"int32\").tofile(f)\n",
    "            if new:\n",
    "                np.array([self.N_valid,\n",
    "                          self.N_invalid,\n",
    "                          self.num_iter],\n",
    "                         dtype=\"int32\").tofile(f)\n",
    "                np.array([self.theta,\n",
    "                          self.lambda_h,\n",
    "                          self.lambda_J,\n",
    "                          self.lambda_group,\n",
    "                          self.N_eff],\n",
    "                         dtype=precision).tofile(f)\n",
    "                self.alphabet.dtype = \"S1\"\n",
    "                self.alphabet[np.where(self.alphabet != b\"\")].tofile(f)\n",
    "                self.alphabet.dtype = \"U1\"\n",
    "                self.weights.astype(precision).tofile(f)\n",
    "            self.target_seq.dtype = \"S1\"\n",
    "            self.target_seq[np.where(self.target_seq != b\"\")].tofile(f)\n",
    "            self.target_seq.dtype = \"U1\"\n",
    "            self.index_list.astype(\"int32\").tofile(f)\n",
    "            self.f_i.astype(precision).tofile(f)\n",
    "            self.h_i.astype(precision).tofile(f)\n",
    "\n",
    "            if not new:\n",
    "                for i in range(self.L - 1):\n",
    "                    for j in range(i + 1, self.L):\n",
    "                        np.array([i + 1, j + 1], dtype=\"int32\").tofile(f)\n",
    "                        self.f_ij[i, j].astype(precision).tofile(f)\n",
    "                        self.J_ij[i, j].astype(precision).tofile(f)\n",
    "            else:\n",
    "                for i in range(self.L - 1):\n",
    "                    for j in range(i + 1, self.L):\n",
    "                        self.f_ij[i, j].astype(precision).tofile(f)\n",
    "\n",
    "                for i in range(self.L - 1):\n",
    "                    for j in range(i + 1, self.L):\n",
    "                        self.J_ij[i, j].astype(precision).tofile(f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Model parameters and mutation effects (EVmutation)\n",
    "\n",
    "## Content\n",
    "\n",
    "This notebook demonstrates how to explore the parameters of undirected graphical models inferred using EVcouplings, and how to use those to quantitatively predict the effects of mutations.\n",
    "\n",
    "## Reference\n",
    "\n",
    "Hopf, T. A., Ingraham, J. B., Poelwijk, F.J., Schrfe, C.P.I., Springer, M., Sander, C., & Marks, D. S. (2017). Mutation effects predicted from sequence co-variation. *Nature Biotechnology* **35**, 128135 doi:10.1038/nbt.3769\n",
    "\n",
    "## Tutorial\n",
    "\n",
    "### Part 1: Load the model parameters from file\n",
    "\n",
    "This file is generated by the couplings stage of the pipeline (using plmc) and has the extension .model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from evcouplings.couplings import CouplingsModel\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load parameters from file to create a pairwise model\n",
    "fn = '/home/as974/ada/mutations/piv_model_results/YAP1_HUMAN_hmmerbit_plmc_n5_m30_f50_t0.2_r170-205_id100_b18.model'\n",
    "#c_1 = CouplingsModel_1(fn,file_format='plmc_v1')\n",
    "c_1 = CouplingsModel_1(fn,file_format='plmc_v1')\n",
    "#c = CouplingsModel(\"/home/as974/ada/persistent-vi/potts/pvi/example/DHFR/PF00186.eij\",file_format='plmc_v1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/n/groups/marks/software/anaconda_o2/envs/evcouplings_stable/lib/python3.6/site-packages/ipykernel_launcher.py:760: RuntimeWarning: invalid value encountered in log\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[ 0.        , 13.14162122,  4.8471668 , ...,  0.4961845 ,\n",
       "         2.15196612,  2.29422328],\n",
       "       [13.14162122,  0.        ,  0.47787534, ...,  2.08846943,\n",
       "         2.16910302,  0.44095644],\n",
       "       [ 4.8471668 ,  0.47787534,  0.        , ...,  2.21974912,\n",
       "         2.16955981,  0.50873214],\n",
       "       ...,\n",
       "       [ 0.4961845 ,  2.08846943,  2.21974912, ...,  0.        ,\n",
       "         0.37063513,  2.30407744],\n",
       "       [ 2.15196612,  2.16910302,  2.16955981, ...,  0.37063513,\n",
       "         0.        ,  2.33302098],\n",
       "       [ 2.29422328,  0.44095644,  0.50873214, ...,  2.30407744,\n",
       "         2.33302098,  0.        ]])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(c_1.fn())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['-', '-', '-', '-', '-', '-', '-', '-', '-', '-', '-', '-', '-',\n",
       "       '-', '-', '-', '-', '-', '-', '-', '-', '-', '-', '-', '-', '-',\n",
       "       '-', '-', '-', '-', '-', '-', '-', '-', '-', '-'], dtype='<U1')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c_1.target_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 2: Predict mutation effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from evcouplings.mutate import predict_mutation_table, single_mutant_matrix\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the experimental mutational scanning dataset for PABP by Melamed et al., RNA, 2013\n",
    "data = pd.read_csv(\n",
    "    \"example/PABP_YEAST_Fields2013-singles.csv\", sep=\";\", comment=\"#\"\n",
    ")\n",
    "\n",
    "# predict mutations using our model\n",
    "data_pred = predict_mutation_table(\n",
    "    c, data, \"effect_prediction_epistatic\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(c_1.double_mut_mat- c.double_mut_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for *: 'NoneType' and 'float'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-df6805e1a38d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# can also add predictions by the corresponding independent model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mc0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mc_1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_independent_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# data_pred = predict_mutation_table(\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#     c0, data_pred, \"effect_prediction_independent\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-4-7cbb10008ac1>\u001b[0m in \u001b[0;36mto_independent_model\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    877\u001b[0m                 \u001b[0m_log_post\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_gradient\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    878\u001b[0m                 \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf_i\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlambda_h\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mN_eff\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 879\u001b[0;31m                 \u001b[0mdisp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    880\u001b[0m             )\n\u001b[1;32m    881\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/n/groups/marks/software/anaconda_o2/envs/evcouplings_stable/lib/python3.6/site-packages/scipy/optimize/optimize.py\u001b[0m in \u001b[0;36mfmin_bfgs\u001b[0;34m(f, x0, fprime, args, gtol, norm, epsilon, maxiter, full_output, disp, retall, callback)\u001b[0m\n\u001b[1;32m    887\u001b[0m             'return_all': retall}\n\u001b[1;32m    888\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 889\u001b[0;31m     \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_minimize_bfgs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfprime\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallback\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallback\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mopts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    890\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    891\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mfull_output\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/n/groups/marks/software/anaconda_o2/envs/evcouplings_stable/lib/python3.6/site-packages/scipy/optimize/optimize.py\u001b[0m in \u001b[0;36m_minimize_bfgs\u001b[0;34m(fun, x0, args, jac, callback, gtol, norm, eps, maxiter, disp, return_all, **unknown_options)\u001b[0m\n\u001b[1;32m    941\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    942\u001b[0m         \u001b[0mgrad_calls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmyfprime\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mwrap_function\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfprime\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 943\u001b[0;31m     \u001b[0mgfk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmyfprime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    944\u001b[0m     \u001b[0mk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    945\u001b[0m     \u001b[0mN\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/n/groups/marks/software/anaconda_o2/envs/evcouplings_stable/lib/python3.6/site-packages/scipy/optimize/optimize.py\u001b[0m in \u001b[0;36mfunction_wrapper\u001b[0;34m(*wrapper_args)\u001b[0m\n\u001b[1;32m    290\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mwrapper_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    291\u001b[0m         \u001b[0mncalls\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 292\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwrapper_args\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    293\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    294\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mncalls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-4-7cbb10008ac1>\u001b[0m in \u001b[0;36m_gradient\u001b[0;34m(x, *args)\u001b[0m\n\u001b[1;32m    868\u001b[0m             \u001b[0mZ\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    869\u001b[0m             \u001b[0mP\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0mZ\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 870\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mN\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mP\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mfi\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mlambda_h\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    871\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    872\u001b[0m         \u001b[0mh_i\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mzeros\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mL\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnum_symbols\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: unsupported operand type(s) for *: 'NoneType' and 'float'"
     ]
    }
   ],
   "source": [
    "# can also add predictions by the corresponding independent model\n",
    "c0 = c_1.to_independent_model()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutant</th>\n",
       "      <th>linear</th>\n",
       "      <th>log</th>\n",
       "      <th>effect_prediction_epistatic</th>\n",
       "      <th>effect_prediction_independent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G126A</td>\n",
       "      <td>0.711743</td>\n",
       "      <td>-0.490571</td>\n",
       "      <td>-2.610615</td>\n",
       "      <td>0.406487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G126C</td>\n",
       "      <td>0.449027</td>\n",
       "      <td>-1.155127</td>\n",
       "      <td>-5.663638</td>\n",
       "      <td>-0.027602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G126E</td>\n",
       "      <td>0.588928</td>\n",
       "      <td>-0.763836</td>\n",
       "      <td>-6.611062</td>\n",
       "      <td>-1.827570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G126D</td>\n",
       "      <td>0.229853</td>\n",
       "      <td>-2.121218</td>\n",
       "      <td>-7.270577</td>\n",
       "      <td>-1.180076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G126N</td>\n",
       "      <td>0.679435</td>\n",
       "      <td>-0.557593</td>\n",
       "      <td>-5.809167</td>\n",
       "      <td>0.387440</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutant    linear       log  effect_prediction_epistatic  \\\n",
       "0  G126A  0.711743 -0.490571                    -2.610615   \n",
       "1  G126C  0.449027 -1.155127                    -5.663638   \n",
       "2  G126E  0.588928 -0.763836                    -6.611062   \n",
       "3  G126D  0.229853 -2.121218                    -7.270577   \n",
       "4  G126N  0.679435 -0.557593                    -5.809167   \n",
       "\n",
       "   effect_prediction_independent  \n",
       "0                       0.406487  \n",
       "1                      -0.027602  \n",
       "2                      -1.827570  \n",
       "3                      -1.180076  \n",
       "4                       0.387440  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_pred.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Predicting single-substitution landscape (independent of experiment) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'single_mutant_matrix' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-a2f052f6eed7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m singles = single_mutant_matrix(\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mc_1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput_column\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"effect_prediction_epistatic\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m )\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0msingles\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'single_mutant_matrix' is not defined"
     ]
    }
   ],
   "source": [
    "singles = single_mutant_matrix(\n",
    "    c_1, output_column=\"effect_prediction_epistatic\"\n",
    ")\n",
    "\n",
    "singles.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "singles = single_mutant_matrix(\n",
    "    c, output_column=\"effect_prediction_epistatic\"\n",
    ")\n",
    "\n",
    "singles.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(c.single_mut_mat_full-c_1.single_mut_mat_full)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Predicting arbitary mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Illegal substitution: L186M\nAlphabet: {'-': 0, 'A': 1, 'C': 2, 'D': 3, 'E': 4, 'F': 5, 'G': 6, 'H': 7, 'I': 8, 'K': 9, 'L': 10, 'M': 11, 'N': 12, 'P': 13, 'Q': 14, 'R': 15, 'S': 16, 'T': 17, 'V': 18, 'W': 19, 'Y': 20}\nPositions: [  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18\n  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36\n  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54\n  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72\n  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90\n  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108\n 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126\n 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144\n 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161]",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/n/groups/marks/software/anaconda_o2/envs/evcouplings_stable/lib/python3.6/site-packages/evcouplings/couplings/model.py\u001b[0m in \u001b[0;36mdelta_hamiltonian\u001b[0;34m(self, substitutions, verify_mutants)\u001b[0m\n\u001b[1;32m    687\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0msubs_pos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubs_from\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubs_to\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubstitutions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 688\u001b[0;31m                 \u001b[0mpos\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_map\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msubs_pos\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    689\u001b[0m                 \u001b[0msubs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malphabet_map\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msubs_to\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 186",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-58-250173b702c8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# double mutant L186M, G188A\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdelta_E\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelta_E_couplings\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelta_E_fields\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdelta_hamiltonian\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m186\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"L\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"M\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m188\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"G\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"A\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mdelta_E\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/n/groups/marks/software/anaconda_o2/envs/evcouplings_stable/lib/python3.6/site-packages/evcouplings/couplings/model.py\u001b[0m in \u001b[0;36mdelta_hamiltonian\u001b[0;34m(self, substitutions, verify_mutants)\u001b[0m\n\u001b[1;32m    697\u001b[0m             raise ValueError(\n\u001b[1;32m    698\u001b[0m                 \"Illegal substitution: {}{}{}\\nAlphabet: {}\\nPositions: {}\".format(\n\u001b[0;32m--> 699\u001b[0;31m                     \u001b[0msubs_from\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubs_pos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubs_to\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malphabet_map\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_list\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    700\u001b[0m                 )\n\u001b[1;32m    701\u001b[0m             )\n",
      "\u001b[0;31mValueError\u001b[0m: Illegal substitution: L186M\nAlphabet: {'-': 0, 'A': 1, 'C': 2, 'D': 3, 'E': 4, 'F': 5, 'G': 6, 'H': 7, 'I': 8, 'K': 9, 'L': 10, 'M': 11, 'N': 12, 'P': 13, 'Q': 14, 'R': 15, 'S': 16, 'T': 17, 'V': 18, 'W': 19, 'Y': 20}\nPositions: [  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18\n  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36\n  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54\n  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72\n  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90\n  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108\n 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126\n 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144\n 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161]"
     ]
    }
   ],
   "source": [
    "# double mutant L186M, G188A\n",
    "delta_E, delta_E_couplings, delta_E_fields = c.delta_hamiltonian([(186, \"L\", \"M\"), (188, \"G\", \"A\")])\n",
    "delta_E"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Short-cuts to single and double mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1.3010851466961952"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# single mutation matrix\n",
    "c.smm(127, \"E\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1.3010851466961952"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c_1.smm(127, \"E\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "186",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-70-70218e5ff3f7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# double mutation matrix (double mutant L186M, G188A)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdmm\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m186\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m188\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"M\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"A\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/n/groups/marks/software/anaconda_o2/envs/evcouplings_stable/lib/python3.6/site-packages/evcouplings/couplings/model.py\u001b[0m in \u001b[0;36mdmm\u001b[0;34m(self, i, j, A_i, A_j)\u001b[0m\n\u001b[1;32m   1187\u001b[0m             substitutions).\n\u001b[1;32m   1188\u001b[0m         \"\"\"\n\u001b[0;32m-> 1189\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__4d_access\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdouble_mut_mat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mA_i\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mA_j\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1190\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1191\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mto_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprecision\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"float32\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile_format\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"plmc_v2\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/n/groups/marks/software/anaconda_o2/envs/evcouplings_stable/lib/python3.6/site-packages/evcouplings/couplings/model.py\u001b[0m in \u001b[0;36m__4d_access\u001b[0;34m(self, matrix, i, j, A_i, A_j)\u001b[0m\n\u001b[1;32m    970\u001b[0m             \u001b[0;36m4\u001b[0m\u001b[0mD\u001b[0m \u001b[0mmatrix\u001b[0m \u001b[0;34m\"matrix\"\u001b[0m \u001b[0msliced\u001b[0m \u001b[0maccording\u001b[0m \u001b[0mto\u001b[0m \u001b[0mvalues\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mA_i\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mA_j\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    971\u001b[0m         \"\"\"\n\u001b[0;32m--> 972\u001b[0;31m         \u001b[0mi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_map\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0m_SLICE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    973\u001b[0m         \u001b[0mj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_map\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0m_SLICE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    974\u001b[0m         \u001b[0mA_i\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mA_i\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malphabet_map\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mA_i\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0m_SLICE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/n/groups/marks/software/anaconda_o2/envs/evcouplings_stable/lib/python3.6/site-packages/evcouplings/couplings/model.py\u001b[0m in \u001b[0;36m__map\u001b[0;34m(self, indices, mapping)\u001b[0m\n\u001b[1;32m    947\u001b[0m             )\n\u001b[1;32m    948\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 949\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mmapping\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindices\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    950\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    951\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__4d_access\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmatrix\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mA_i\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mA_j\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 186"
     ]
    }
   ],
   "source": [
    "# double mutation matrix (double mutant L186M, G188A)\n",
    "c.dmm(186, 188, \"M\", \"A\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Statistical energy of a sequence (rather than delta to WT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69.8888109875413"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "E, E_couplings, E_fields = c_1.hamiltonians([c_1.seq()])[0]\n",
    "E"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69.8888109875413"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "E, E_couplings, E_fields = c.hamiltonians([c.seq()])[0]\n",
    "E"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 3: Explore model parameters\n",
    "\n",
    "Please see the documentation of evcouplings.couplings.model.CouplingsModel for all available methods. Note that most of these methods can be accessed using lists of positions/symbols. \n",
    "\n",
    "Examples include:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-'\n",
      " '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-'\n",
      " '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-' '-']\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "127",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-44-618f41af3e1e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# symbol for particular position (or list of positions)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc_1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m127\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-4-7cbb10008ac1>\u001b[0m in \u001b[0;36mseq\u001b[0;34m(self, i)\u001b[0m\n\u001b[1;32m   1098\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtarget_seq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1099\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1100\u001b[0;31m             \u001b[0mi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_map\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1101\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtarget_seq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1102\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-4-7cbb10008ac1>\u001b[0m in \u001b[0;36m__map\u001b[0;34m(self, indices, mapping)\u001b[0m\n\u001b[1;32m    915\u001b[0m             )\n\u001b[1;32m    916\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 917\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mmapping\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindices\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    918\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    919\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__4d_access\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmatrix\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mA_i\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mA_j\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 127"
     ]
    }
   ],
   "source": [
    "# full sequence\n",
    "print(c_1.seq())\n",
    "\n",
    "# symbol for particular position (or list of positions)\n",
    "print(c_1.seq(127))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140\n",
      " 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158\n",
      " 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176\n",
      " 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194\n",
      " 195 196 197 198 199 200 201 202 203 204]\n"
     ]
    }
   ],
   "source": [
    "# positions in model\n",
    "print(c.index_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A' 'C' 'D' 'E' 'F' 'G' 'H' 'I' 'K' 'L' 'M' 'N' 'P' 'Q' 'R' 'S' 'T' 'V'\n",
      " 'W' 'Y']\n"
     ]
    }
   ],
   "source": [
    "# alphabet\n",
    "print(c.alphabet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.2060956209897995"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pair coupling parameters\n",
    "c.Jij(127, 172, c.seq(127), c.seq(172))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.30619758"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# field parameters\n",
    "c.hi(127, c.seq(127))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 4: Index mapping for complexes\n",
    "\n",
    "When inferring parameters for a concatenated sequence alignment (e.g. protein complexes or other discontinuous sequence segments), the internal model numbering does not match the numbering of the actual sequence. In this case, the model numbering can be manually remapped such that the model can be indexed using tuples (segment_id, position_in_segment).\n",
    "\n",
    "*(Note this example does not execute here and is intended as a reference of how to use the code only)*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from evcouplings.couplings import Segment, SegmentIndexMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define which segments the concatenated sequence in the model is made of.\n",
    "# Most important parameters are region_start and region_end of Segment\n",
    "s_pare = Segment(\"aa\", \"F7YBW7\", 1, 103, segment_id=\"parE\")\n",
    "s_pard = Segment(\"aa\", \"F7YBW8\", 1, 93, segment_id=\"parD\")\n",
    "\n",
    "# Create index mapper\n",
    "mapper = SegmentIndexMapper(True, 1, s_pare, s_pard)\n",
    "\n",
    "# Update indices in model to complex numbering\n",
    "c_mapped = mapper.patch_model(c, inplace=False)\n",
    "\n",
    "# Now access model using tuples of indices rather than single positions alone\n",
    "print(c.seq((\"parD\", 59)))\n",
    "print(c.smm((\"parD\", 59), \"A\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
